### ISSUE TITLE
Please add serde derives for Serialize/Deserialize

### ISSUE THREAD

**Issue by RReverser** [stbuehler/rust-cidr#2]
Currently it's hard to serialize/deserialize network structs without using hacks described in https://serde.rs/remote-derive.html. Adding `#[derive(Serialize, Deserialize)]` directly on types would help a lot.

**Comment by stbuehler** [comment]
Pulling `serde-derive` is quite heavy - but I guess it could be an optional feature?

**Comment by stbuehler** [comment]
Also given how the struct fields are private, deriving `Serialize` and `Deserialize` seems like a bad idea: if I ever change the internal representation, the serialized data is probably not compatible anymore.

Which means I'd need to decide on some format - either string representation, or some documented "struct"-like serializiation.

Any suggestions? :)

**Comment by RReverser** [comment]
> but I guess it could be an optional feature

Yup, that should work.

> Which means I'd need to decide on some format - either string representation, or some documented "struct"-like serializiation.

I'd suggest to do what serde does for IpAddr types - see https://docs.serde.rs/src/serde/ser/impls.rs.html#529-542.

TL;DR - if you have human-readable serializer, you might serialize standard string representation like `1.2.3.4/12` and deserialize using existing FromStr, and if it's binary format, write out octets (this is already implemented on IpAddr, so you just need to call `.serialize`) followed by a single byte for network length.

**Comment by RReverser** [comment]
Alternatively, just use current field names and add tests to make sure representation stays the same; `address` and `network_length` seem pretty good :)

**Comment by stbuehler** [comment]
I implemented it manually, and made it an optional feature (but enabled by default; `serde` without `serde-derive` is not that heavy).

Feedback would be nice :)

**Comment by stbuehler** [comment]
Pushed a variant using a different schema to a separate branch.

For those`Serializer`s using newtype information I still add it (I guess `pickle` might be using it), but otherwise the data is compatible over all cidr/inet types (as long as the value is in the required subset), and should use a very space efficient representation (but doesn't drop unused octets from the cidr values).

**Comment by stbuehler** [comment]
Though it seems upstream prefers something similar to what `serde-derive` would generate, they don't see a problem with other variants.

So I'll use the second approach (tuple instead of struct/map/...), mainly because the code is much shorter.  I'll keep the newtype wrappers for now unless someone gives me a good reason to drop them :)

**Comment by RReverser** [comment]
One thing I had in mind but didn't explain when I commented is ability to use `serialize_bytes` for IPv4/IPv6 addresses which I used in the shared gist.

All these static arrays are serialized as `&[u8]` so they still encode length, and then `&[u8]` by default has suboptimal treatment in serde serializers/deserializers due to lack of specialisation in stable Rust. This is a known issue in serde, and they even have a special crate to work around that on type level - https://github.com/serde-rs/bytes.

This is why I suggested to just use `serialize_bytes` (or wrap using that crate wrapper), in which case actual raw bytes will be optimally encoded in non-human serializers, and you will also always have length encoded too, and can use that to distinguish between IPv4 and IPv6 instead of yet another custom tag like I did in gist.

**Comment by stbuehler** [comment]
> All these static arrays are serialized as &[u8] so they still encode length

No, that is not true.  The arrays `[u8; 4]` and `[u8; 16]` are serialized as tuples of length 4 or 16 (see https://docs.rs/serde/1.0.27/src/serde/ser/impls.rs.html#137).  Tuple lengths may be stored by some formats, but you need to pass the length again to `deserialize_tuple` - storing tuple lengths is only interesting if either your format absolutely requires it or you just store it as variable-length sequence.

Sure, performance could perhaps be better with specialized impls, but the format doesn't depend on it, and it might get fixed in the future.  It's also possible the optimizer sees through all of it and inlines to something like memcpy...

**Comment by RReverser** [comment]
> It's also possible the optimizer sees through all of it and inlines to something like memcpy...

It's not about that - serializing/deserializing 4 or even 16 bytes is fast enough, it's more about space efficiency of representation.

**Comment by RReverser** [comment]
Here I made simple gist comparing resulting current and tuple+bytes-based representations: https://gist.github.com/RReverser/7f4caf4d6aeef3713d05bfad2e3c012e

Results against popular binary serializers such as `bincode` and `cbor`:
```
bincode
bincode
  Current implementation:      [10 00 00 00 00 00 00 00 31 39 32 2E 31 36 38 2E 31 30 30 2E 30 2F 32 34]
  Length-based implementation: [04 00 00 00 00 00 00 00 C0 A8 64 00 18]
cbor
  Current implementation:      [82 18 18 84 18 C0 18 A8 18 64 00]
  Length-based implementation: [82 44 C0 A8 64 00 18 18]
```

You can see how it can save almost 2x space in some cases, not to mention simpler (and so faster) encoding/decoding.

**Comment by RReverser** [comment]
(UPD: added MessagePack to the comparison)

**Comment by stbuehler** [comment]
The current `binpack` release (0.9.2) still flags itself as "human readable".  With [`serde_test::Compact`](https://docs.rs/serde_test/1.0.27/serde_test/struct.Compact.html) you can disable that, and then the results look exactly like I predicted them.  Also the `binpack` 1.0.0 release will fix that.

Also see 86167ea53b8ef1d335f9f8c398b7d2158db65e1e.

`MessagePack` again uses human readable encoding; I guess with `serde_test::Compact` it would look similar to the `cbor` output.

Both `cbor` and `MessagePack` want to preserve the data schema, and I'm not trying to optimize for them; you're basically asking me to `binpack` my data, and then serialize the result as bytestring to `cbor` and `MessagePack`.

You also can clearly see how much space your implementation is wasting with a format like `binpack` that actually tries to optimize for space.
